lib/llvmopencl/VariableUniformityAnalysis.cc

```diff
@@ -279,7 +279,9 @@ bool VariableUniformityAnalysisResult::isUniform(llvm::Function *F,
     return true;
   }
 
-  if (isa<llvm::ConstantInt>(V)) {
+  // We cannot check for Constant here as the magic id variables are external
+  // constants.
+  if (isa<llvm::ConstantData>(V)) {
     setUniform(F, V, true);
```

Comment:
Do you mean `GlobalValue`s by external constants? Why they can't be marked uniform? They are constants and therefore they must be uniform (their address not their contained value).

Reply:
Yes, in general, but not in this case because the ID pseudo variables are modeled as global variables as this point, and can thus diverge.

Reply:
> ID pseudo variables are modeled as global variables as this point, and can thus diverge.

You can still mark them uniform. What should be marked possibly divergent are loads from them. 

```
@some_global_var = external global i32  ;  @some_global_var is uniform.

%1 = load i32, ptr @some_global_var  ; %1 could be divergent.
```

Reply:
Yes, that should work. Currently it's recursively done on operands until the leaf and this would add a special case at load handling.  I'll check if it can be done without too intrusive changes.